Documentation: warning on multi-threading/processing (#5287)
pyodide/pyodide: docs/usage/faq.md

    print(js_lambda, js_lambda_, js_lambda__) # 7, 7, 8
`);
```

> ## Can i use multi-threading/processing?

Reviewer: ```suggestion
## Can I use multi-threading/processing?
```

> ## Can I use multi-threading/processing?

Reviewer 2: ```suggestion
## Can I use threading/multiprocessing?
```

Due to the limitation of current [wasm](https://pyodide.org/en/stable/usage/wasm-constraints.html)

> all packages that uses parallelism (eg. rely on _fork_ and _pthread_)

Reviewer: ```suggestion
all packages that use parallelism (eg. rely on _fork_ and _pthread_)
```

> Due to the limitation of current [wasm](https://pyodide.org/en/stable/usage/wasm-constraints.html)
> all packages that use parallelism (eg. rely on _fork_ and _pthread_)
> **may not work**. Accepting the performance degradation, a workaround
> consists of locking the number of threads down to `1` like shown below:

Reviewer 2: ```suggestion
No, fork and pthreads do not work in Pyodide. Attempts to use `threading`, `multiprocessing`, or `subprocess` will raise an `OSError`. You may be able to work around this by setting the number of threads to 1:
```

Reviewer: Should we also mention that any attempts to use their APIs will cause fatal errors but simply `import threading`, etc. should not cause problems?

Author: > Should we also mention that any attempts to use their APIs will cause fatal errors but simply `import threading`, etc. should not cause problems?

You're right. General info APIs and synchronous  work on the main thread, but I don't know how reliable the results are. For example this works:

```pycon
>>> current = threading.current_thread()
>>> current.name
'MainThread'
>>> current.is_alive()
True
>>> import threading
>>> def target (nums):
...     print(sum(nums))
... 
>>> t = threading.Thread(target=target, args = ([1,2,3],))
>>> t.run()
6
>>>
```

Reviewer 2: > any attempts to use their APIs will cause fatal errors

I don't believe that anything causes fatal errors. Just normal Python `OSError`s.

Reviewer: Yes

Reviewer: Maybe it's still better to guard the imports (or bring them into the functions where they are used) anyway.

```py
is_wasm = sys.platform == "emscripten" or platform.machine() in ["wasm32", "wasm64"]

if is_wasm:
  # set n_threads = 1
```

> Ultimately, developers will need to add specific support for the platform.

Reviewer 2: I don't think this sentence adds anything.
```suggestion
```

    print(js_lambda, js_lambda_, js_lambda__) # 7, 7, 8
`);
```

## Can I use threading/multiprocessing/subprocess?

No, fork and pthreads do not work in Pyodide (see more [here](https://pyodide.org/en/stable/usage/wasm-constraints.html)).
Attempts to use `threading`, `multiprocessing`, or `subprocess` will raise a `RuntimeError`.
You may be able to work around this by setting the number of threads to 1:

```py

> def _can_start_thread() -> bool:

Reviewer 2: For some reason the `->` is marked as red here in the diff UI. I thought it meant there was a non ascii character but it seems to be normal...

Reviewer 2: Do you see the red thing? Maybe it's a github bug? I think no action is needed.

Author: Yes i can see the red arrow as well

Reviewer 2: _shrug_

    if sys.platform == "emscripten":
        return sys._emscripten_info.pthreads
    return platform.machine() not in ("wasm32", "wasm64")

can_start_thread = _can_start_thread()

if not can_start_thread:
  # set n_threads = 1
```

> Bear in mind that you can still import the packages and use the general info API, but you cannot

Reviewer 2: ```suggestion
You can still import the packages and use the general info API, but you cannot
```

start any asynchronous work without receiving a `RuntimeError`.

```pycon
>>> import threading
>>> current = threading.current_thread()
>>> current.name
'MainThread'
>>> current.daemon
False
>>> current.is_alive()
True
>>> def target(nums):
...     print(sum(nums))
...
>>> t = threading.Thread(target=target, args=([1, 2, 3], ))
>>> t.run()
6
>>> t.start()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python312.zip/threading.py", line 994, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread

> >>>

Reviewer 2: Maybe delete this line:
```suggestion
```
